### ISSUE TITLE
Nil checks discrepancy between building directly or through require

### ISSUE THREAD

**Issue by Daniel-Worrall** [crystal-lang/crystal#9926]
The [code builds](https://carc.in/#/r/9yre), but not when required from a different file

file1.cr
```cr
array = [1,2,3] 
 
1.times do
  number = array.find { |num| num == 1 }
  if number
    [number]
  end
end
```

file2.cr
```cr
require "./file1"
```

Error:
```
Showing last frame. Use --error-trace for full trace.

In file1.cr:6:4

 6 | [number]
     ^-
Error: no overload matches 'Pointer(Int32)#[]=' with types Int32, (Int32 | Nil)

Overloads are:
 - Pointer(T)#[]=(offset, value : T)
Couldn't find overloads for these types:
 - Pointer(Int32)#[]=(offset : Int32, value : Nil)
```

**Comment by HertzDevil** [comment]
I am unable to replicate this on 1.3.2.

**Comment by Daniel-Worrall** [comment]
This was reproduced on 0.35.1, not on 0.36.0 and above.
Closing since fixed

**Comment by Daniel-Worrall** [comment]
Bisected, this was fixed in #9986
